plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/ComposeMachineProviderImpl.java

```diff
@@ -0,0 +1,619 @@
+
+            DockerNode node = dockerMachineFactory.createNode(workspaceId, container);
+            if (isDev) {
+                node.bindWorkspace();
+                LOG.info("Machine with id '{}' backed by container '{}' has been deployed on node '{}'",
+                         machineId, container, node.getHost());
+            }
+
+            dockerInstanceStopDetector.startDetection(container,
+                                                      machineId,
+                                                      workspaceId);
+
+            MachineImpl machine = new MachineImpl(MachineConfigImpl.builder()
+                                                                   .setDev(isDev)
+                                                                   .setName(machineName)
+                                                                   .setType("docker")
+                                                                   // casting considered as safe because more than int of megabytes is a lot!
+                                                                   .setLimits(new LimitsImpl((int)Size
+                                                                           .parseSizeToMegabytes(
+                                                                                   service.getMemLimit() + "b")))
+                                                                   .setSource(new MachineSourceImpl(service.getBuild() != null ?
+                                                                                                    "context" :
+                                                                                                    "image")
+                                                                                      .setLocation(service.getBuild() != null ?
+                                                                                                   service.getBuild().getContext() :
+                                                                                                   service.getImage()))
+                                                                   .build(),
+                                                  machineId,
+                                                  workspaceId,
+                                                  envName,
+                                                  namespace,
+                                                  MachineStatus.RUNNING,
+                                                  null);
+
+            return dockerMachineFactory.createInstance(machine,
+                                                       container,
+                                                       image,
+                                                       node,
+                                                       machineLogger);
+        } catch (SourceNotFoundException e) {
+            throw e;
```

Comment:
it is needed. As you can see in that case SourceNotFoundException will be re-thrown, otherwise any exception will be wrapped with ServerException